src/math_tools/math_tools.hpp

```diff
@@ -89,16 +91,14 @@ inline std::size_t factorial(std::size_t f)
     return r;
 }
 
-template <class T, std::size_t D>
-KOKKOS_INLINE_FUNCTION T dot_product(std::array<T, D> const& a, std::array<T, D> const& b)
+template <class T, class... Dims>
+KOKKOS_INLINE_FUNCTION T
+dot_product(Vector<T, Dims...> const& a, Vector<T, typename Dims::Dual...> const& b)
```

Comment:
Should this one be moved to the directory containing tensor.hpp in the future ?

We could make a distinction between math functions that launch a kernel from the ones that execute in the calling thread ?

Reply:
> Should this one be moved to the directory containing tensor.hpp in the future ?

I don't think so. The directory containing tensor.hpp contains data types. This is a mathematical function that happens to use a particular data type.

> We could make a distinction between math functions that launch a kernel from the ones that execute in the calling thread ?

I agree that this could be useful. I think this file should only contain functions that execute in the calling thread

Reply:
I mean we chose to define math operators (`operator+` and others) in `tensor.hpp`, the distinction with `dot_product` is not clear to me.

Reply:
Ah good point